sound/soc/sof/sof-audio.c

```diff
@@ -75,12 +67,37 @@ static int sof_dai_config_setup(struct snd_sof_dev *sdev, struct snd_sof_dai *da
 	return ret;
 }
 
+static int sof_widget_kcontrol_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
+{
+	struct snd_sof_control *scontrol;
+	int ret;
+
+	/* set up all controls for the widget */
+	list_for_each_entry(scontrol, &sdev->kcontrol_list, list)
+		if (scontrol->comp_id == swidget->comp_id) {
+			ret = sof_kcontrol_setup(sdev, scontrol);
+			if (ret < 0) {
+				dev_err(sdev->dev, "error: fail to set up kcontrols for widget %s\n",
+					swidget->widget->name);
+				return ret;
+			}
+		}
+
+	return 0;
+}
+
 static int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 {
 	struct sof_ipc_free ipc_free;
 	struct sof_ipc_reply reply;
 	int ret;
 
+	/* only free when refcount is 0 */
+	if (!atomic_dec_and_test(&swidget->ref_count))
+		return 0;
```

Comment:
I might be wrong, but this still seems racy to me. Let's say the test passes, i.e. you only have 1 user currently active, but right here after the test another user opens it, wouldn't that be possible? I heard in a comment, that `atomic_t` is tricky - it's rarely a sufficient atomicity primitive. Maybe you need to lock the use-counter together with actual adding to / removing from the list here?

Reply:
@lyakh  The PCM hw_params and hw_free are protected with the card->pcm_mutex already. The only place I can imagine a race is with system suspend but even there hw_free() will be called before the PCM is suspended and restarted upon resume.